### ISSUE TITLE
How to determine if Wayland backend is used [Nvidia]

### ISSUE THREAD

**Issue by cnekmp** [qtile/qtile#3297]
After upgrading Nvidia drivers I've lanuched Qtile from tty with the following command `exec qtile start -b wayland` and it launched without any problems. 
The question is how can I be sure that I'm currently using Qtile with Wayland backend?

`Nvidia Driver Version: 510.47.03`

```
[farid@unknown ~]$ qtile -v
0.16.2.dev901+gda78ea40
[farid@unknown ~]$ pacman -Qi qtile
Name            : qtile-git
Version         : 0.20.0.r19.g7c51e130-1
Description     : A full-featured, pure-Python tiling window manager. (git version)
Architecture    : x86_64
URL             : http://www.qtile.org
Licenses        : MIT
Groups          : None
Provides        : qtile
Depends On      : gdk-pixbuf2  glibc  pango  python-cairocffi  python-cffi  python-xcffib
Optional Deps   : alsa-utils: volume widget [installed]
                  canto-daemon: canto widget
                  cmus: cmus widget
                  jupyter_console: interaction with qtile via Jupyter
                  khal: khal_calendar widget
                  libpulse: for pulse_volume and pulseaudio_ffi widget [installed]
                  librsvg: for SVG support in some widgets widgets or wallpapers [installed]
                  lm_sensors: sensors widget [installed]
                  moc: moc widget
                  python-dbus-next: for utils, notifications and several widgets [installed]
                  python-iwlib: wlan widget [installed]
                  python-keyring: imapwidget widget
                  python-mpd2: mpd2widget widget
                  python-psutil: graph, net and memory widget [installed]
                  python-pywlroots: Wayland backend [installed]
                  xorg-xwayland: Wayland backend [installed]
                  python-setproctitle: change process name to qtile
                  python-xdg: launchbar widget [installed]
Required By     : None
Optional For    : None
Conflicts With  : qtile
Replaces        : None
Installed Size  : 3.65 MiB
Packager        : Unknown Packager
Build Date      : Fri 04 Feb 2022 11:49:48 AM +04
Install Date    : Fri 04 Feb 2022 11:50:00 AM +04
Install Reason  : Explicitly installed
Install Script  : Yes
Validated By    : None
```

Thanks

**Comment by elParaguayo** [comment]
If you pass the `-b wayland` argument then I'm not sure it's possible for you to be using the x11 backend!

However, for peace of mind, run `qtile cmd-obj -o cmd -f eval -a "self.core.name"` and you should see `(True, 'wayland')`

**Comment by cnekmp** [comment]
> Is there any reason you think you're not using wayland?

Yes. Because everything works flawlessly. And seems I was right, Qtile was launched via X11 backend. Souldn't Qtrile throw an error in case it cannot launch wayland backend instead of switching to X11?

Once again, I would like to remind that I've run `exec qtile start -b wayland` from `tty` directly. And here is the output:

```
[farid@unknown ~]$ qtile cmd-obj -o cmd -f eval -a "self.core.name"
(True, 'x11')
[farid@unknown ~]$
```

**Comment by elParaguayo** [comment]
Wow. Didn't expect that! I'm not sure how that's possible.

If you're starting from tty directly. You don't need `exec`. Can you see if that makes a difference?

**Comment by cnekmp** [comment]
Without `exec` I get following error:
```
[farid@unknown ~]$ qtile start -b wayland
python: render/wlr_renderer.c:240: wlr_renderer_init_wl_shm: Assertion `argb8888 && xrgb8888' failed.
Aborted (core dumped)
```

![IMG_20220204_135414](https://user-images.githubusercontent.com/35117097/152509548-72669c5c-dbb9-42bd-b156-2f7083605dba.jpg)

**Comment by elParaguayo** [comment]
Right. Well at least that's trying to load the wayland backend now. I've only ever seen `exec` used in `xinit` related scripts but I don't fully understand it's behaviour.

Might need @m-col to comment on that new error as he's the Wayland guru.

**Comment by m-col** [comment]
>  I've only ever seen exec used in xinit related scripts but I don't fully understand it's behaviour.

`exec` replaces the current process (i.e. the shell) with the process you're `exec`ing. In a TTY this would mean that once `qtile` exits, the shell exits, and you're back to the log-in prompt. I can see why this would be desirable - when your session ends (and if it crashes unexpectedly) an open TTY isn't available. It doesn't swallow arguments - if you run `exec qtile cmd-obj -o cmd -f eval -a "self.core.name"` in a terminal you might see the output for a moment but your shell exits and your terminal probably will close :)

> Once again, I would like to remind that I've run exec qtile start -b wayland from tty directly. And here is the output:

The X11 backend cannot start up if it cannot connect to an X server. One must have been started somewhere on the system, and maybe it finds it and runs as the WM in that. Not sure.

> python: render/wlr_renderer.c:240: wlr_renderer_init_wl_shm: Assertion `argb8888 && xrgb8888' failed.

So I have some bad news about the nvidia driver and wlroots support, see: https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/3349. There isn't anything we can do on our end, I'm afraid, except hold out hope that nvidia makes any necessary changes. For a sliver of hope, see this sway commit message: https://github.com/swaywm/sway/commit/b48cb6b0ec1320ad25fd2c0a1b5118dbe2536060

**Comment by elParaguayo** [comment]
Help me out. If `exec` doesn't swallow arguments, how have they managed to start an `x11` backend when running the command `exec qtile start -b wayland`. This tells qtile to use the wayland backend modules, I don't think there's any discovery/fallback handling in there so I don't see how this ends up with x11.

**Comment by m-col** [comment]
Ah I am misremembering, if `os.environ.get("DISPLAY")` is falsey then Qtile exits. The DISPLAY variable must have been set?

**Comment by elParaguayo** [comment]
Don't think that changes it either. Anyway, maybe not a conversation for this thread...

**Comment by cnekmp** [comment]
Thanks @m-col for links and explanation. But I think that it would be nice to restrict/prevent `-b wayland` option to launch Qtile with X11 backend if it cannot launch Wayland. I mean someone may think that he's already running wayland backend without issues and have an effect of placebo, while he's running X11 in background :)

**Comment by m-col** [comment]
That's the bit I don't quite understand. There is no code path that could make Qtile start with a backend other than the one specified, with the exception that the default is "x11": https://github.com/qtile/qtile/blob/master/libqtile/scripts/start.py#L51. If you pass `-b wayland` then it tries to start the wayland core, and it fails, and the x11 core is never even imported. Just so I'm on the same page is this what you are seeing?:

1. start from fresh TTY
2. run `exec qtile start -b wayland`
3. A graphical session starts
4. Inside the session, `qtile cmd-obj -o cmd -f eval -a "self.core.name"` prints `(True, 'x11')`

**Comment by cnekmp** [comment]
> * start from fresh TTY
> * run `exec qtile start -b wayland`
> * A graphical session starts
> * Inside the session, `qtile cmd-obj -o cmd -f eval -a "self.core.name"` prints `(True, 'x11')`

Yes. This is what exactly happens in my case. While I didn't set any exports in my bashrc or profile regarding X11.

**Comment by m-col** [comment]
Lol what, that doesn't make any sense at all! Please could you post the output of `env` both from the TTY before starting it, and from the terminal you used for the `qtile cmd-obj` call?

**Comment by cnekmp** [comment]
env in tty:

```
SHELL=/bin/bash
JAVA_HOME=/usr/lib/jvm/java-17-openjdk
EDITOR=vim
XDG_SEAT=seat0
DEFAULT_JAVA=java-17-openjdk
PWD=/home/farid
LOGNAME=farid
XDG_SESSION_TYPE=tty
MOTD_SHOWN=pam
HOME=/home/farid
LANG=en_US.UTF-8
XDG_SESSION_CLASS=user
TERM=xterm
USER=farid
MANPAGER=sh -c 'col -bx | bat -l man -p'
SHLVL=1
XDG_VTNR=1
XDG_SESSION_ID=3
SYSTEMD_LESS=FRXMK
LD_LIBRARY_PATH=:/usr/local/lib
XDG_RUNTIME_DIR=/run/user/1000
BROWSER=google-chrome-stable
PATH=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/farid/.local/bin
SAL_USE_VCLPLUGIN=gtk3
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
MAIL=/var/spool/mail/farid
_=/usr/bin/env

```

env in Qtile (alacritty):
```
SHELL=/bin/bash
WINDOWID=16777218
COLORTERM=truecolor
JAVA_HOME=/usr/lib/jvm/java-17-openjdk
EDITOR=vim
GTK_MODULES=canberra-gtk-module
XDG_SEAT=seat0
DEFAULT_JAVA=java-17-openjdk
PWD=/home/farid
LOGNAME=farid
XDG_SESSION_TYPE=tty
XAUTHORITY=/home/farid/.Xauthority
WINDOWPATH=1
MOTD_SHOWN=pam
HOME=/home/farid
LANG=en_US.UTF-8
ALACRITTY_SOCKET=/run/user/1000/Alacritty-:0-811778.sock
XDG_SESSION_CLASS=user
TERM=xterm
USER=farid
MANPAGER=sh -c 'col -bx | bat -l man -p'
DISPLAY=:0
SHLVL=3
XDG_VTNR=1
XDG_SESSION_ID=3
SYSTEMD_LESS=FRXMK
LD_LIBRARY_PATH=:/usr/local/lib:/usr/local/lib
XDG_RUNTIME_DIR=/run/user/1000
BROWSER=google-chrome-stable
PATH=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/farid/.local/bin:/home/farid/.local/bin
ALACRITTY_LOG=/tmp/Alacritty-811778.log
SAL_USE_VCLPLUGIN=gtk3
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
MAIL=/var/spool/mail/farid
_=/usr/bin/env
```

https://user-images.githubusercontent.com/35117097/152658840-a34b504f-4b55-4c33-8fb7-ecdc688bc881.mp4

**Comment by cnekmp** [comment]
Ok, I've found the root cause of issue. Turns out, in case if you have `exec qtile start` in your `.xinitrc` (like in my case), running command `exec qtile start -b wayland` will launch X11 backend. But if you'll edit your `.xinitrc` and comment out or remove `exec qtile start` line and try again, it won't launch and throw back to tty with error. I don't know how `.xinitrc` may affect exec command from shell but seems it does.

**Comment by m-col** [comment]
Check the top of the screen: 
![image](https://user-images.githubusercontent.com/44536113/152661536-597e62e8-3d1a-4e15-a46f-575c6cc9f92d.png)

It's just logged in. Which means that the `exec qtile start -b wayland` fails, exits immediately, jumps you back to the TTY login, but you have an auto-login service enabled, and that's auto-starting X, which starts qtile via your `.xinitrc`. If you just do `qtile start -b wayland` without `exec` it will exit without restarting the TTY.

**Comment by cnekmp** [comment]
@m-col one question, wlroots just updated on Arch and now I get different kind of error when I try to launch wayland backend. Should I open another issue for this case?

```
[farid@unknown ~]$ qtile start -b wayland
Traceback (most recent call last):
  File "/usr/bin/qtile", line 33, in <module>
    sys.exit(load_entry_point('qtile==0.20.1.dev26+g909e2a64', 'console_scripts', 'qtile')())
  File "/home/farid/.local/lib/python3.10/site-packages/libqtile/scripts/main.py", line 66, in main
    options.func(options)
  File "/home/farid/.local/lib/python3.10/site-packages/libqtile/scripts/start.py", line 90, in start
    q = make_qtile(options)
  File "/home/farid/.local/lib/python3.10/site-packages/libqtile/scripts/start.py", line 51, in make_qtile
    kore = libqtile.backend.get_core(options.backend)
  File "/home/farid/.local/lib/python3.10/site-packages/libqtile/backend/__init__.py", line 15, in get_core
    return importlib.import_module(f"libqtile.backend.{backend}.core").Core(*args)
  File "/home/farid/.local/lib/python3.10/site-packages/libqtile/backend/wayland/core.py", line 93, in __init__
    self.compositor, self.backend = wlroots_helper.build_compositor(self.display)
ValueError: too many values to unpack (expected 2)
```

```
[farid@unknown ~]$ p -Qi wlroots
Name            : wlroots
Version         : 0.15.1-2
Description     : Modular Wayland compositor library
Architecture    : x86_64
URL             : https://gitlab.freedesktop.org/wlroots/wlroots
Licenses        : MIT
Groups          : None
Provides        : libwlroots.so=10-64
Depends On      : libglvnd  libinput  seatd  libseat.so=1-64  libxcb  libxkbcommon  libxkbcommon.so=0-64
                  opengl-driver  pixman  libpixman-1.so=0-64  systemd-libs  libudev.so=1-64
                  vulkan-icd-loader  libvulkan.so=1-64  wayland  xcb-util-errors  xcb-util-renderutil
                  xcb-util-wm
Optional Deps   : xorg-xwayland: Xwayland support [installed]
Required By     : python-pywlroots
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 1118.48 KiB
Packager        : Brett Cornwall <brett@i--b.com>
Build Date      : Mon 07 Feb 2022 06:48:58 AM +04
Install Date    : Mon 07 Feb 2022 04:35:58 PM +04
Install Reason  : Installed as a dependency for another package
Install Script  : No
Validated By    : Signature
```

**Comment by m-col** [comment]
I think you need to update `pywlroots`

**Comment by cnekmp** [comment]
> I think you need to update `pywlroots`

Already done. You think this is still because of Nvidia?

```
[farid@unknown ~]$ p -Qi pywlroots
Name            : python-pywlroots
Version         : 0.15.7-2
Description     : Python binding to the wlroots library using cffi
Architecture    : x86_64
URL             : https://github.com/flacjacket/pywlroots
Licenses        : custom:UIUC
Groups          : None
Provides        : pywlroots
Depends On      : glibc  python-cffi  python-pywayland  python-xkbcommon  wlroots  libwlroots.so=10-64
Optional Deps   : None
Required By     : None
Optional For    : qtile-git
Conflicts With  : pywlroots
Replaces        : pywlroots
Installed Size  : 1203.15 KiB
Packager        : David Runge <dvzrv@archlinux.org>
Build Date      : Fri 04 Feb 2022 12:52:26 AM +04
Install Date    : Fri 04 Feb 2022 11:47:00 AM +04
Install Reason  : Installed as a dependency for another package
Install Script  : No
Validated By    : Signature

[farid@unknown ~]$ paru
no new news
:: Synchronizing package databases...
 core is up to date
 extra is up to date
 community is up to date
 multilib is up to date
:: Starting full system upgrade...
 there is nothing to do
:: Looking for AUR upgrades
:: Looking for devel upgrades
 there is nothing to do
[farid@unknown ~]$
```

**Comment by m-col** [comment]
It looks like you have `/usr/bin/qtile`, a system-wide installation, and a local user installation at `~/.local`, and these are in conflict. I think you're invoking the system-wide installation but it is importing outdated modules from `~/.local`.

**Comment by cnekmp** [comment]
> It looks like you have `/usr/bin/qtile`, a system-wide installation, and a local user installation at `~/.local`, and these are in conflict. I think you're invoking the system-wide installation but it is importing outdated modules from `~/.local`.

That was the case! Uninstalled local installation with `pip uninstall qtile` and now it loads Wayland backend correctly :) Although, there are some issues with (no mouse pointer, some glitches, rofi is not running correctly etc.). At least, it runs :)

Waiting for Nvidia, Wayland, XWayland to be stable in a future )

**Comment by m-col** [comment]
There are some glitches and xwayland probllems currently but your mouse pointer should be there! That could well be an nvidia thing though...

**Comment by cnekmp** [comment]
> There are some glitches and xwayland probllems currently but your mouse pointer should be there! That could well be an nvidia thing though...

Cursor is ok by default. But when I move it on the XWayland launched window (browser, rofi, etc.) it disappears visually until I logout
